### DISCUSSION TITLE
Implement Speculation Rules API in prefetch

### DISCUSSION THREAD

**Discussion by rossrobino** [withastro/roadmap#794] [Proposal]
Chrome currently has the [speculation rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API) enabled without a flag. This api allows you to pre-render multipage applications/html on the client, even running the script tags on the page. This helps for a faster experience and to reduce layout shift. It'd be great if Astro prefetch had an experimental option to enable this progressive enhancement. 

Here's a prefetch function I've wrote to do this, when it's supported.

https://github.com/rossrobino/domco/blob/ee9ab7926a22f4e88d2971e3061801a5056105b7/packages/domco/helpers/prefetch/index.ts#L71

I really enjoy using Astro, thanks for your consideration!

**Comment by bluwy** [comment]
I'm open to supporting this in `prefetch`! The reason I didn't look too deep into this before was because of its browser support, and I didn't want to add extra code for a feature that won't work everywhere.

Adding as a new experimental option is a good start for me, feel free to contribute a PR if you like.

**Reply by rossrobino** [reply]
Ok great! I will try and do that!

**Reply by bluwy** [reply]
Awesome. As a hint, this is the file you'll need to update: https://github.com/withastro/astro/blob/5b4e41a86ff54bf17aeb4516c155e4b9cd9a7a1b/packages/astro/src/prefetch/index.ts

Line 16-18 has some examples you can use to retrieve the experimental option value so that it can be treeshaken away if the feature is not enabled.

**Reply by rossrobino** [reply]
Got it, appreciate it! I'll put something together

**Reply by rossrobino** [reply]
Submitted a PR: https://github.com/withastro/astro/pull/9644
